lib/bom.gradle

```diff
@@ -21,7 +21,27 @@ configure(projectsWithFlags('bom')) {
                 }
                 return "${a.ext.artifactId}".compareTo("${b.ext.artifactId}")
             }.each { p ->
-                api "${p.group}:${p.ext.artifactId}:${p.version}"
+                if (rootProject.ext.has('bomGroups')) {
+                    def bomGroups = rootProject.ext.bomGroups
+                    if (!(bomGroups instanceof Map)) {
+                        throw new IllegalStateException("bomGroups must be a Map: ${bomGroups}")
+                    }
+
+                    if (!bomGroups.containsKey(project.path)) {
+                        api "${p.group}:${p.ext.artifactId}:${p.version}"
+                    } else {
+                        def subs = bomGroups.get(project.path)
+                        if (!(subs.value instanceof List)) {
+                            throw new IllegalStateException("bomGroups' value must be a List: ${subs.value}")
+                        }
+
+                        if (subs.contains(p.path)) {
+                            api "${p.group}:${p.ext.artifactId}:${p.version}"
+                        }
```

Comment:
Should we raise an `IllegalStateException` if a `publish` project is not defined in a `bomGroups`?
For example, `module1:submodule1` to `module1:submodule3` is flagged with `publish` and only `module1:submodule1` and `module1:submodule2` are set in `bomGroups`.
if `bomGroups` is defined, I'd like to assume that all `publish` + `java` projects should belong to one of `bomGroups`.

Reply:
> For example, ``module1:submodule1` to `module1:submodule3` is flagged with publish and only `module1:submodule1` and `module1:submodule2` are set in bomGroups.

In this case, `module1:submodule1` is only included.

> if bomGroups is defined, I'd like to assume that all publish + java projects should belong to one of bomGroups

I'm sorry. I don't understand this. Could you explain more with examples?

Reply:
Let's say that we have the following example without  `bomGroups` definition.
```
includeWithFlags ':bom', 'bom'
includeWithFlags ':module1', 'java', 'publish'
includeWithFlags ':module2', 'java', 'publish'
includeWithFlags ':moduleN', 'java', 'publish'
```
All modules flagged by `publish` are automatically added to the BOM.

However, when `bomGroups` is used, from what I understand, users need to update `bomGroups` whenever a new `publish` project is added.
If they missed updating `bomGroups`, should we raise an exception to notify the misconfiguration? 
```
ext {
    bomGroups = [
            ':module1': [':module1:submodule1', ':module1:submodule2'],
            ':module2': [':module2:submodule1', ':module2:submodule2']
    ]
}

// root project
includeWithFlags ':module1', 'bom'
includeWithFlags ':module1:submodule1', 'java', 'publish'
includeWithFlags ':module1:submodule2', 'java', 'publish'

includeWithFlags ':module2', 'bom'
includeWithFlags ':module2:submodule1', 'java', 'publish'
includeWithFlags ':module2:submodule2', 'java', 'publish'
includeWithFlags ':module2:submodule3', 'java', 'publish' // <- Which bom blongs to?
```

Reply:
@ikhoon 
I'm sorry for late answer.

> However, when bomGroups is used, from what I understand, users need to update bomGroups whenever a new publish project is added.
If they missed updating bomGroups, should we raise an exception to notify the misconfiguration?

Certainly.. Umm, I think that you want to make `bomGroup` be restricted. It is a good option to prevent misconfiguration. However, sometimes, a user doesn't want to include a project which has a `publish` flag. We can't distinguish `won't include a project to bom` and `misconfiguration`.

If you'd like to make `bomGroup` include all `submodules` to `module`, we don't need `bomGroups` property. It is similar to my first PR https://github.com/line/gradle-scripts/pull/113.

I'd like to keep the rules flexible. What do you think?

Reply:
> I'd like to keep the rule flexible. When there isn't some submodule in bomGroups, can we print a warning log?

I see. I'm also OK with leaving a warning log.